[1.12] Continue loading mod recipes and advancements after encountering an error

Choonster opened the pull request:
This fixes #4013.

Currently, `CraftingHelper.findFiles` exits after the `processor` function returns `false`;
which means that it stops loading recipes or advancements for a mod as soon as one of them has a syntax error or is otherwise invalid.

This PR changes `CraftingHelper.findFiles` to always call `processor` for every file and only return `true` if every call to `processor` returned `true`.

`CraftingHelper.findFiles` still exits early under every other failure condition (couldn't find Minecraft JAR, couldn't get filesystem iterator, `preprocessor` returned `false`), as it did previously.

To demonstrate this, I've added a recipe that should be loaded after the `conditions_property_not_array` recipe. The syntax exception in `conditions_property_not_array` would previously prevent the new recipe from loading, but the changes made in this PR allow the new recipe to be loaded and used in-game.

Choonster committed:
Continue loading mod recipes and advancements after encountering an error

bs2609 commented on src/main/java/net/minecraftforge/common/crafting/CraftingHelper.java:
```diff
@@ -803,13 +805,12 @@ else if (source.isDirectory())
-                    if (cont == null || !cont.booleanValue())
-                        return false;
+                    Boolean ret = processor.apply(root, itr.next());
+                    success &= ret == null ? false : ret;
```
`ret == null ? false : ret` can be replaced with `ret != null && ret`, which is a little more consise.

Choonster replied:
Fixed, thanks.

Choonster committed:
Simplify boolean expression

mezz commented:
Looks good to me. Tested and confirmed it works as intended.

LexManos commented:
Seems to me like this should be something thats opted into by the caller.

Choonster committed:
Make visiting all files opt-in behaviour for CraftingHelper.findFiles

Choonster commented:
I've added a new overload of `CraftingHelper.findFiles` with a parameter that controls whether or not all files are visited. I've also deprecated the old overloads and switched the recipe and advancement loaders to use the new overload.
